dask/array/routines.py

```diff
@@ -882,6 +885,293 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
+          if necessary. This may induce a runtime increase.
+    bins : sequence of arrays describing bin edges, int, or sequence of ints
+        The bin specification.
+
+        The possible binning configurations are:
+
+        * A sequence of arrays describing the monotonically increasing
+          bin edges along each dimension.
+        * A single int describing the total number of bins that will
+          be used in each dimension (this requires the ``range``
+          argument to be defined).
+        * A sequence of ints describing the total number of bins to be
+          used in each dimension (this requires the ``range`` argument
+          to be defined).
+
+        When bins are described by arrays, the rightmost edge is
+        included. Bins described by arrays also allows for non-uniform
+        bin widths.
+    range : sequence of pairs, optional
+        A sequence of length D, each a (min, max) tuple giving the
+        outer bin edges to be used if the edges are not given
+        explicitly in ``bins``. If defined, this argument is required
+        to have an entry for each dimension. Unlike
+        :func:`numpy.histogramdd`, if `bins` does not define bin
+        edges, this argument is required (this function will not
+        automatically use the min and max of of the value in a given
+        dimension because the input data may be lazy in dask).
+    normed : bool, optional
+        An alias for the density argument that behaves identically. To
+        avoid confusion with the broken argument to `histogram`,
+        `density` should be preferred.
+    weights : dask Array, optional
+        An array of values weighing each sample in the input data. The
+        chunks of the the weights must be identical to the chunking
+        along the 0th axis of the data sample.
+    density : bool, optional
+        If ``False`` (default), the returned array represents the
+        number of samples in each bin. If ``True``, the returned array
+        represents the probability density function at each bin.
+
+    See Also
+    --------
+    histogram
+
+    Examples
+    --------
+    Computing the histogram in 5 blocks using different bin edges
+    along each dimension:
+
+    >>> import dask.array as da
+    >>> x = da.random.uniform(0, 1, size=(1000, 3), chunks=(200, 3))
+    >>> edges = [
+    ...     np.linspace(0, 1, 5), # 4 bins in 1st dim
+    ...     np.linspace(0, 1, 6), # 5 in the 2nd
+    ...     np.linspace(0, 1, 4), # 3 in the 3rd
+    ... ]
+    >>> h, edges = da.histogramdd(x, bins=edges)
+    >>> result = h.compute()
+    >>> result.shape
+    (4, 5, 3)
+
+    Defining the bins by total number and their ranges, along with
+    using weights:
+
+    >>> bins = (4, 5, 3)
+    >>> ranges = ((0, 1),) * 3  # expands to ((0, 1), (0, 1), (0, 1))
+    >>> w = da.random.uniform(0, 1, size=(1000,), chunks=x.chunksize[0])
+    >>> h, edges = da.histogramdd(x, bins=bins, range=ranges, weights=w)
```

Comment:
Please excuse  the beginner question: I already find this a bit confusing for `da.histogram` when dealing with `dask.DataFrame`s. How would this look like if my input values are in a DataFrame? Essentially, I put for all arrays a `df["value_column"].to_dask_array()`?

I found in `da.histogram` that using this works as long as no weights are passed:
```py
h, bins = da.histogram(df["value_column"], bins=50,
    range=[-8.0e-23, 8.0e-23])
```

If I pass `weights=df["weights"]` from my `weights` column in the DataFrame, I got:
```
AttributeError: 'Series' object has no attribute 'chunks'
```
(`Series` is `dask.dataframe.core.Series`)

For 1D, I am using this (hope that is the right way to use it):
```py
h, bins = da.histogram(df["value_column"].to_dask_array(), bins=50,
    range=[-8.0e-23, 8.0e-23],
    weights=df["weights"].to_dask_array())
```

Reply:
My `df` is a `dd.from_delayed(dfs)` with `dfs` being a list of delayed pandas dataframes, similar to what is documented [here](https://docs.dask.org/en/latest/delayed-collections.html).

Ref.: https://github.com/openPMD/openPMD-api/pull/935

Reply:
For the 1D case:
If weights are included in the function call then there is a check to see if the chunks of the values array and the chunks of the weights array are compatible. If weights are absent, the Series works because `values.chunks` is never called (and the chunks property/attr only exists for a `dask.array.Array`). The `dask.dataframe.Series` will be just be passed along to the new graph and treated like an arbitrary dask collection. Perhaps this is something that should be addressed in the docs!

For the nD case:
Right now my implementation is using `sample.chunks` to make sure we don't have data chunked in a way that creates a boundary between individual coordinates (see the table in the docstring). Again, something that I'll add to the docs. Maybe we can try to detect DataFrame/Series objects and call `to_dask_array` in the function body, I'll ask what @jakirkham thinks about that.

Reply:
What you propose Doug makes sense to me 🙂

Reply:
After reading a bit more of the `dask.array` module I think it would be more consistent with the rest of dask to avoid calling `to_dask_array` in the `histogramdd` function body. Making it clear in the docstring seems to be the best thing to do.

Reply:
Yep that makes sense. Going to mark this as resolved then